hedera-node/src/main/java/com/hedera/services/store/EntityStore.java

```diff
@@ -0,0 +1,286 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.hedera.services.context.properties.GlobalDynamicProperties;
+import com.hedera.services.exceptions.InvalidTransactionException;
+import com.hedera.services.records.TransactionRecordService;
+import com.hedera.services.state.merkle.MerkleAccount;
+import com.hedera.services.state.merkle.MerkleEntityAssociation;
+import com.hedera.services.state.merkle.MerkleEntityId;
+import com.hedera.services.state.merkle.MerkleToken;
+import com.hedera.services.state.merkle.MerkleTokenRelStatus;
+import com.hedera.services.state.submerkle.EntityId;
+import com.hedera.services.store.models.Account;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.Token;
+import com.hedera.services.store.models.TokenRelationship;
+import com.hedera.services.txns.validation.OptionValidator;
+import com.swirlds.fcmap.FCMap;
+import org.apache.commons.lang3.NotImplementedException;
+
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+import static com.hedera.services.exceptions.ValidationUtils.validateFalse;
+import static com.hedera.services.exceptions.ValidationUtils.validateTrue;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.ACCOUNT_DELETED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.ACCOUNT_EXPIRED_AND_PENDING_REMOVAL;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_ACCOUNT_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_NOT_ASSOCIATED_TO_ACCOUNT;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_WAS_DELETED;
+
+/**
+ * Loads and saves entities to and from the Swirlds state, hiding the details of
+ * Merkle types from client code by providing an interface in terms of model
+ * objects whose methods can perform validated business logic.
+ *
+ * When loading an entity, fails fast by throwing an {@link InvalidTransactionException}
+ * if the entity is not usable in normal business logic. There are three such
+ * cases:
+ * <ol>
+ *     <li>The entity is missing.</li>
+ *     <li>The entity is deleted.</li>
+ *     <li>The entity is expired and pending removal.</li>
+ * </ol>
+ * Note that in the third case, there <i>is</i> one valid use of the entity;
+ * namely, in an update transaction whose only purpose is to manually renew
+ * the expired entity. Such update transactions must use a dedicated
+ * expiry-extension service, which will be implemented before TokenUpdate.
+ *
+ * When saving an entity, invites an injected {@link TransactionRecordService} to
+ * inspect the entity for changes that may need to be included in the record
+ * of the transaction.
+ */
+public class EntityStore {
```

Comment:
If we are going to modularize with modules for Contract Service, Consensus Service, Token Service, File Service, etc., then instead of EntityStore, we probably want to have a different store for each type of service. So in this case, I think TokenStore is still the right name for tokens, but accounts might live in some other store. Along the same lines, I don't think that com.hedera.servcies.store is the right package for TokenStore to live in. For the current non-modular organization, EntityStore in this package is probably the right choice, but I wonder whether we should create modular-friendly packages now and start putting things in them so that as we go over the code and refactor to fit this new design we are also putting things in the right places, and then we can modularize as the next step. Thoughts?

Reply:
Split `EntityStore` into `AccountStore` and `TypedTokenStore` to facilitate future steps toward modularity.